lib/vm/src/vmoffsets.rs

```diff
@@ -112,8 +144,111 @@ impl VMOffsets {
         self.num_local_memories = cast_to_u32(module.memories.len());
         self.num_local_globals = cast_to_u32(module.globals.len());
         self.has_trap_handlers = true;
+        self.precompute();
         self
     }
+
+    /// Number of local tables defined in the module
+    pub fn num_local_tables(&self) -> u32 {
+        self.num_local_tables
+    }
+
+    /// Number of local memories defined in the module
+    pub fn num_local_memories(&self) -> u32 {
+        self.num_local_memories
+    }
+
+    fn precompute(&mut self) {
+        self.vmctx_signature_ids_begin = 0;
+        self.vmctx_imported_functions_begin = self
```

Comment:
This is so much more comprehensible than the previous code already, but I do wonder if there’s an opportunity to further simplify this code with some helpers.

For instance: instead of manually having `checked_add` and checked_mul`, having that in a utility function along the lines of `fma` (fused multiply-add).

That way we’d be looking at something like

```rust
self.vmctx_imported_functions_begin = offset_fma(
    self.num_signature_ids, 
    u32::from(self.size_of_vmshared_signature_index())
    self.vmctx_signature_ids_begin
);
```

which seems like less noise overall. (names/argument ordering at your choice)

Reply:
Great idea thanks! I’ve just pushed a commit doing this change :)